pkg/services/pluginsintegration/keyretriever/dynamic/dynamic_retriever.go

```diff
@@ -0,0 +1,230 @@
+package dynamic
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/log"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/keyretriever/static"
+	// Only used for getting the feature flag value
+	// "github.com/grafana/grafana/pkg/services/featuremgmt"
+)
+
+const publicKeySyncInterval = 10 * 24 * time.Hour // 10 days
+
+// ManifestKeys is the database representation of public keys
+// used to verify plugin manifests.
+type ManifestKeys struct {
+	KeyID     string `json:"keyId"`
+	PublicKey string `json:"public"`
+	Since     int64  `json:"since"`
+}
+
+type KeyRetriever struct {
+	cfg *config.Cfg
+	log log.Logger
+
+	lock    sync.Mutex
+	cli     http.Client
+	kv      plugins.KeyStore
+	hasKeys bool
+}
+
+var _ plugins.KeyRetriever = (*KeyRetriever)(nil)
+
+func New(cfg *config.Cfg, kv plugins.KeyStore) *KeyRetriever {
+	kr := &KeyRetriever{
+		cfg: cfg,
+		log: log.New("plugin.signature.key_retriever"),
+		cli: makeHttpClient(),
+		kv:  kv,
+	}
+	return kr
+}
+
+func (kr *KeyRetriever) Run(ctx context.Context) error {
```

Comment:
Nit: I'm not 100% on this, but will call it out as it confused me a little at first. I think maybe it's better to rename this something slightly different like `Exec`/`Execute` so as not to wrongly assume `KeyRetriever` itself is a background service

Reply:
Take a look to this alternative. I am now registering this `dynamic_retriever` as the background service (which is disabled depending on the feature flag). The `keyretriever` wrapper is now a normal service, that receives a `dynamic_retriever` as a dependency but only uses it if the feature flag is set. WDYT, better or worse?

Reply:
Yep I like this better!